### ISSUE TITLE
How to use this code for super-resolution?

### ISSUE THREAD

**Issue by Lyken17** [jcjohnson/fast-neural-style#63]
Hi author, I noticed that there is a 'task' option in your code, as mentioned in your paper, this network can also be used for super-resolution task. How can I modify this code to do upsample task?

**Comment by htoyryla** [comment]
I have been thinking of this, too. I guess one would need a dataset consisting of hires images and their downscaled copies, and then one would need a network architecture that outputs an upscaled image. 

I was thinking that because there is the 'upsample' parameter, maybe all that would be needed would be hires images in the dataset, and they would be scaled down during training. Looking at DataLoader.lua this doesn't seem to be the case.

Maybe all that is needed is a new script to construct a dataset of hires images and their downscaled copies. And an arch that produces an upsampled image, but that should be easy to configure.

**Comment by jcjohnson** [comment]
You need to do three things to train a super-resolution network:

(1) Use a different preprocessing script to prepare and HDF5 dataset of low-resolution and high-resolution image patches: https://gist.github.com/jcjohnson/97c9f9d73c66ae87174a14e7ea7198fb

(2) When running train.lua, set the `-task` flag to `upsample`; this changes the way the DataLoader works, so it will return batches of HR and LR patches.

(3) Set up an architecture (using the `-arch` flag in train.lua) whose output is larger than its input.

Once you have a model trained, `fast_neural_style.lua` should let you apply a super-resolution model to images or directories of images. I found that the raw outputs from the networks sometimes had strange color shifts, so as a postprocessing step I did histogram matching between the raw CNN outputs and the low-res inputs: https://gist.github.com/jcjohnson/e01e4fcf7b7dfa9e0dbee6c53d3120b6

I'll eventually fold this code and info into the main repo; I should have more time to do so in a few weeks.

**Comment by charlesdlp** [comment]
@jcjohnson 
Hi Justin, I am trying to train new models on a p2.8xlarge, but I am available to train only one model at a time. Do you know how can I train different models at the same time on this GPU ?

Charles

**Comment by jcjohnson** [comment]
@charlesdlp Set the environment variable `CUDA_VISIBLE_DEVICES` before starting `train.lua`; for example to run on GPU 4:

```bash
export CUDA_VISIBLE_DEVICES=4
th train.lua # Other args here
```

**Comment by charlesdlp** [comment]
Dear Justin,

Thank you really much for your help ! Let me try that settings.

Best

Charles

**Comment by ffallentan** [comment]
@jcjohnson 
Hi Justin, I've used the make_sr_dataset.py with `--sizes 4` to prepare an HDF5 dataset, so I can train a p4xlarge model. However, when I use the train.lua with `-task upsample -arch c9s1-64,R64,R64,R64,R64,u64,u64,c9s1-3` according to the supplementary material mentioned in your paper, I got the following error.
```
/home/avtest/torch/install/share/lua/5.1/cudnn/init.lua:126: Error in CuDNN: CUDNN_STATUS_BAD_PARAM (cudnnGetConvolutionNdForwardOutputDim)
stack traceback:
	[C]: in function 'error'
	/home/avtest/torch/install/share/lua/5.1/cudnn/init.lua:126: in function 'errcheck'
	...torch/install/share/lua/5.1/cudnn/SpatialConvolution.lua:135: in function 'createIODescriptors'
	...torch/install/share/lua/5.1/cudnn/SpatialConvolution.lua:366: in function <...torch/install/share/lua/5.1/cudnn/SpatialConvolution.lua:363>
	[C]: in function 'xpcall'
	/home/avtest/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/home/avtest/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function </home/avtest/torch/install/share/lua/5.1/nn/Sequential.lua:41>
	[C]: in function 'xpcall'
	/home/avtest/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/home/avtest/torch/install/share/lua/5.1/nn/ConcatTable.lua:11: in function </home/avtest/torch/install/share/lua/5.1/nn/ConcatTable.lua:9>
	...
	[C]: in function 'xpcall'
	/home/avtest/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/home/avtest/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	train.lua:174: in function 'opfunc'
	/home/avtest/torch/install/share/lua/5.1/optim/adam.lua:37: in function 'adam'
	train.lua:239: in function 'main'
	train.lua:327: in main chunk
	[C]: in function 'dofile'
	...test/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00405d50
```
I've also tried to set `--sizes 4 --patch_size 288` to generate the dataset matching the size mentioned in your paper, but the `train.lua` throws another error. 

Is there something wrong with my settings? Could you please take a look? Thanks a lot!

**Comment by Lyken17** [comment]
@ffallentan 
What is your CUDA and CUDNN version?  CUDNN_STATUS_BAD_PARAM is usually caused by uncompatible CUDA.  I suggest you to update cuda/cudnn  and run it again.

**Comment by ffallentan** [comment]
@Lyken17 
With the CUDA8.0 and cuDNN v5.1 installed, it throws the same error. 
Could you please paste your settings? Thanks a lot.

**Comment by Lyken17** [comment]
@ffallentan 
I simply use the default setting, and things work well on my server.

I guess it is caused by network parameter. Have you tried default 'arch'?  `cudnnGetConvolutionNdForwardOutputDim` shows the dims of some layers don't fit.

**Comment by ffallentan** [comment]
@Lyken17 
With the default 'arch', the input and output layers have the same dimension, and I think it's not suit for the super-resolution training.

**Comment by Lyken17** [comment]
@ffallentan 
I am not clear what the author did in paper, but I think there is no problem to let input and output have the same dimension when doing super-resolution.

For example, now I have a `400 * 400` RGB image, and want to upscale it to `800 * 800`, what I do is 

1. Scaled the `400 * 400` image to `800 * 800` using the naive method
2. Feed the `3 * 800 * 800` tensor into network, and then let network complete the details.
3. Output is also a `3 * 800 * 800` tensor, and super-resolution task is done.

**Comment by ffallentan** [comment]
@Lyken17 
according to Justin's answer
`(3) Set up an architecture (using the -arch flag in train.lua) whose output is larger than its input.`

**Comment by htoyryla** [comment]
Also setting task to upsample makes the dataloader use different resolution for input and output. It is obvious that this is the way the program is designed to work: upsampling i.e. Increasing size by inserting missing pixels. 

Lykes17's concept might work too, but it is a very different task, modifying existing pixels instead of inserting new pixels. I doubt if it can ever work as well as inserting pixels.

**Comment by ffallentan** [comment]
@htoyryla 
This is my full settings:
`th train.lua -h5_file sr_4x.h5 -checkpoint_name sr_4x -gpu 0 -task upsample -arch c9s1-64,R64,R64,R64,R64,u64,u64,c9s1-3 -padding_type zero`
and throws an error:
`./fast_neural_style/StyleLoss.lua:46: bad argument #1 to 'size' (out of range)`

**Comment by htoyryla** [comment]
22.11.2016, 10:47, ffallentan kirjoitti:
>
> @htoyryla <https://github.com/htoyryla>
> This is my full settings:
> |th train.lua -h5_file sr_4x.h5 -checkpoint_name sr_4x -gpu 0 -task 
> upsample -arch c9s1-64,R64,R64,R64,R64,u64,u64,c9s1-3 -padding_type zero|
> and throws an error:
> |./fast_neural_style/StyleLoss.lua:46: bad argument #1 to 'size' (out 
> of range)|
>
>

I have not yet tried superresolution myself; I'll try when I get home. 
But there is something strange with your error, the line referred to is 
a comment in the present code 
https://github.com/jcjohnson/fast-neural-style/blob/master/fast_neural_style/StyleLoss.lua#L46

Perhaps you should pull the latest code or try with a clean clone.

**Comment by htoyryla** [comment]
22.11.2016, 10:47, ffallentan kirjoitti:
>
> |th train.lua -h5_file sr_4x.h5 -checkpoint_name sr_4x -gpu 0 -task 
> upsample -arch c9s1-64,R64,R64,R64,R64,u64,u64,c9s1-3 -padding_type zero|
> and throws an error:
> |./fast_neural_style/StyleLoss.lua:46: bad argument #1 to 'size' (out 
> of range)|
>
|I am getting the same error, only on line 45 which makes more sense. 
Could be that some changes to the program would still be needed to 
handle upsampling (guessing this because the error happens when 
calculating style loss... why does it even get there?).|

**Comment by ntxfz** [comment]
@ffallentan @htoyryla 
I have met the same problem with you

**./fast_neural_style/StyleLoss.lua:46: bad argument #1 to 'size' (out of range)_**

How did you solve it? Could you please tell me.

**Comment by groot-1313** [comment]
Can someone please shed light on what parameters are to be filled in for the -arch flag?

**Comment by groot-1313** [comment]
For upsampling by a factor of 2, this arch should work fine, right?
`-arch c9s1-64,R64,R64,R64,R64,u64,c9s1-3`

And if it a grayscale image to be upsampled, 
`-arch c9s1-64,R64,R64,R64,R64,u64,u64,c9s1-1`

The above 2 should work well without errors for their suitable application, I would like to clarify if I am right or missing a point.

**Comment by 3DTOPO** [comment]
@htoyryla how did you specify an empty array for the style layers?

Might anyone be willing to share a super-res model just so I can see if it is even worth further investigating training my own model? (2x or 4x would be perfect)

What would a valid architecture be for 4x?

The training command I am running is:
`th train.lua -h5_file superRes_trainSet_284_4x_A.h5 -task upsample   -checkpoint_name upsample_checkpoint_A_5k -gpu 0 -num_iterations 5000 -batch_size 1  -arch c9s1-64,R64,R64,R64,R64,u64,c9s1-3`

But I am getting different errors (using a 4x dataset prepped with make_sr_dataset.py):

`/home/prime/torch/install/bin/luajit: /home/prime/torch/install/share/lua/5.1/nn/Container.lua:67: `
`In 17 module of nn.Sequential:`
`/home/prime/torch/install/share/lua/5.1/nn/THNN.lua:109: inconsistent tensor size at /home/prime/torch/extra/nn/lib/THNN/generic/MSECriterion.c:17`
`stack traceback:`
`	[C]: in function 'v'`
`	/home/prime/torch/install/share/lua/5.1/nn/THNN.lua:109: in function 'MSECriterion_updateOutput' `
`	/home/prime/torch/install/share/lua/5.1/nn/MSECriterion.lua:14: in function 'forward'`
`	./fast_neural_style/ContentLoss.lua:43: in function <./fast_neural_style/ContentLoss.lua:39>`
`	[C]: in function 'xpcall'`
`	/home/prime/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'`
`	/home/prime/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	./fast_neural_style/PerceptualCriterion.lua:134: in function 'forward'`
`	train.lua:197: in function 'opfunc'`
`	/home/prime/torch/install/share/lua/5.1/optim/adam.lua:37: in function 'adam'`
`	train.lua:239: in function 'main'`
`	train.lua:327: in main chunk`
`	[C]: in function 'dofile'`
`	...rime/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00405d50`

`WARNING: If you see a stack trace below, it doesn't point to the place where this error occured. Please use only the one above.`
`stack traceback:`
`	[C]: in function 'error'`
`	/home/prime/torch/install/share/lua/5.1/nn/Container.lua:67: in function 'rethrowErrors'`
`	/home/prime/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'`
`	./fast_neural_style/PerceptualCriterion.lua:134: in function 'forward'`
`	train.lua:197: in function 'opfunc'`
`	/home/prime/torch/install/share/lua/5.1/optim/adam.lua:37: in function 'adam'`
`	train.lua:239: in function 'main'`
`	train.lua:327: in main chunk`
`	[C]: in function 'dofile'`
`	...rime/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk`
`	[C]: at 0x00405d50`
